bigchaindb/core.py

```diff
@@ -49,12 +49,51 @@ def __init__(self, bigchaindb=None):
         self.new_height = None
 
     def init_chain(self, genesis):
-        """Initialize chain with block of height 0"""
-
-        validator_set = [vutils.decode_validator(v) for v in genesis.validators]
-        block = Block(app_hash='', height=0, transactions=[])
+        """Initialize chain upon genesis or a migration"""
+
+        app_hash = ''
+        height = 0
+
+        known_chain = self.bigchaindb.get_latest_abci_chain()
+        if known_chain is not None:
+            chain_id = known_chain['chain_id']
+
+            if known_chain['is_synced']:
+                msg = f'Ignoring the InitChain ABCI request ({genesis}) - ' + \
+                      'the chain {chain_id} is already synced.'
+
+                logger.error(msg)
+                return ResponseInitChain()
+
+            if chain_id != genesis.chain_id:
+                msg = f'Got mismatching chain ID in the InitChain ' + \
+                      'ABCI request - you need to migrate the ABCI client ' + \
+                      'and set new chain ID: {chain_id}.'
+                logger.error(msg)
+                return ResponseInitChain()
```

Comment:
No error codes are documented for `InitChain`.

Reply:
Since we cannot send error codes back we need to panic, just like Tendermint does when it sees a height greater than what it has witnessed.

Reply:
Which goal do we achieve by causing panic?

If there is a process manager, the process will be restarted and it will start all over.

Reply:
The reason I am suggesting to panic is because the system is starting with invalid information. There is not point sending success response when we can't do any further processing.

Reply:
Tendermint state is changed somehow if we return a response, what is not the case if we panic, so you have point.

Reply:
I am in the process of updating other ABCI handlers - it probably makes sense to abort everywhere until the latest chain is in sync, right?